go-controller/pkg/clustermanager/node_network_controller.go

```diff
@@ -0,0 +1,168 @@
+			if err := nnc.clusterSubnetAllocator.MarkSubnetsAllocated(node.Name, hostSubnets...); err != nil {
+				klog.Errorf("Failed to mark the subnet %v as allocated in the cluster subnet allocator for node %s: %w", hostSubnets, node.Name, err)
+			}
+		} else {
+			klog.V(5).Infof("Node %s contains no subnets for network : %s", node.Name, nnc.networkName)
+		}
+	}
+
+	return nil
+}
+
+func (nnc *NodeNetworkController) addUpdateNode(node *corev1.Node) error {
+	nnc.clusterSubnetAllocator.Lock()
+	defer nnc.clusterSubnetAllocator.Unlock()
+
+	existingSubnets, err := util.ParseNodeHostSubnetAnnotation(node, nnc.networkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		// Log the error and try to allocate new subnets
+		klog.Warning("Failed to get node %s host subnets annotations for network %s : %v", node.Name, nnc.networkName, err)
+	}
+
+	hostSubnets, allocatedSubnets, err := nnc.clusterSubnetAllocator.AllocateNodeSubnets(node.Name, existingSubnets, config.IPv4Mode, config.IPv6Mode)
+	if err != nil {
+		return err
+	}
+
+	if len(allocatedSubnets) == 0 {
+		return nil
+	}
+
+	// Release the allocation on error
+	defer func() {
+		if err != nil {
+			if errR := nnc.clusterSubnetAllocator.ReleaseNodeSubnets(node.Name, allocatedSubnets...); errR != nil {
+				klog.Warningf("Error releasing node %s subnets: %v", node.Name, errR)
+			}
+		}
+	}()
+
+	hostSubnetsMap := map[string][]*net.IPNet{nnc.networkName: hostSubnets}
+
+	err = nnc.updateNodeAnnotationWithRetry(node.Name, hostSubnetsMap)
+	return err
+}
+
+func (nnc *NodeNetworkController) deleteNode(node *corev1.Node) error {
+	nnc.clusterSubnetAllocator.Lock()
+	defer nnc.clusterSubnetAllocator.Unlock()
+	nnc.clusterSubnetAllocator.ReleaseAllNodeSubnets(node.Name)
+	return nil
+}
+
+func (nnc *NodeNetworkController) updateNodeAnnotationWithRetry(nodeName string, hostSubnetsMap map[string][]*net.IPNet) error {
+	// Retry if it fails because of potential conflict which is transient. Return error in the
+	// case of other errors (say temporary API server down), and it will be taken care of by the
+	// retry mechanism.
+	resultErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// Informer cache should not be mutated, so get a copy of the object
+		node, err := nnc.watchFactory.GetNode(nodeName)
+		if err != nil {
+			return err
+		}
+
+		cnode := node.DeepCopy()
+		for netName, hostSubnets := range hostSubnetsMap {
+			cnode.Annotations, err = util.UpdateNodeHostSubnetAnnotation(cnode.Annotations, hostSubnets, netName)
+			if err != nil {
+				return fmt.Errorf("failed to update node %q annotation subnet %s",
+					node.Name, util.JoinIPNets(hostSubnets, ","))
+			}
+		}
+		return nnc.kube.UpdateNode(cnode)
+	})
+	if resultErr != nil {
+		return fmt.Errorf("failed to update node %s annotation", nodeName)
+	}
+
+	return nil
+}
+
+func (nnc *NodeNetworkController) cleanup(networkName string) error {
+	if nnc.networkName != networkName {
+		return nil
+	}
```

Comment:
Since the cleanup is called internally from with in the seconday network cluster controller manager I thought having this check.  node network manager is also used by default network cluster controller.

I can remove the check or not take networkName at all since we already store it in the struct nodeNetworkManager.networkName.